### PR TITLE
`strip_path` explicit support, and `x-kong-route-defaults` support

### PR DESCRIPTION
**Pull request by dimitropoulos**
closes: INS-122, closes INS-555, closes INS-554

### CODE REVIEW

File: packages/openapi-2-kong/src/declarative-config/services.js [hunk 1]
```diff
@@ -63,14 +64,18 @@ export function generateService(
...
       const fullPathRegex = pathVariablesToRegex(routePath);
       const route: DCRoute = {
+        ...rootRouteDefaults,
+        ...pathRouteDefaults,
         tags,
         name: generateRouteName(api, routePath, method),
         methods: [method.toUpperCase()],
         paths: [fullPathRegex],
-        strip_path: false,
+        strip_path: api.strip_path || false,
```
Comment (reynolek) [inline packages/openapi-2-kong/src/declarative-config/services.js]: So there are 2 things mentioned in that JIRA ticket -- 1) Provide a way to set `strip_path` to override the default and 2) The [docs](https://www.npmjs.com/package/openapi-2-kong#services) explicitly say that `strip_path` is always true, when clearly in the code its default is false.

I think at this point the best thing to do is to update the [README](https://github.com/Kong/insomnia/blob/develop/packages/openapi-2-kong/README.md#servicesroutes) to say that it defaults to false

File: packages/openapi-2-kong/src/declarative-config/__tests__/services.test.js [hunk 2]
```diff
@@ -97,19 +90,105 @@ describe('services', () => {
...
+            ...baseSpecResult,
+            routes: baseSpecResult.routes.map(route => ({
+              ...route,
+              foo: 'bar',
+            })),
+          },
+        ]);
+      });
+
+      it('will not override options on the route with `x-kong-route-defaults` at the root', async () => {
```
Comment (reynolek) [inline packages/openapi-2-kong/src/declarative-config/__tests__/services.test.js]: Should this be `will override`?

File: packages/openapi-2-kong/src/declarative-config/__tests__/services.test.js [hunk 3]
```diff
@@ -97,19 +90,105 @@ describe('services', () => {
...
+            ...baseSpecResult,
+            routes: baseSpecResult.routes.map(route => ({
+              ...route,
+              ...(route.paths[0] === '/cats$' ? { foo: 'bar' } : {}),
+            })),
+          },
+        ]);
+      });
+
+      it('will not override options on the route with `x-kong-route-defaults` at the root', async () => {
```
Comment (reynolek) [inline packages/openapi-2-kong/src/declarative-config/__tests__/services.test.js]: This test and the one above it have the exact same string description, which makes it hard to know what is different between the two.

File: packages/openapi-2-kong/src/declarative-config/__tests__/services.test.js [hunk 4]
```diff
@@ -97,19 +90,105 @@ describe('services', () => {
...
+            ...baseSpecResult,
+            routes: baseSpecResult.routes.map(route => ({
+              ...route,
+              foo: route.paths[0] === '/cats$' ? 'foobar' : 'bar',
+            })),
+          },
+        ]);
+      });
+
+      it('allows overriding strip_path at the root', async () => {
```
Comment (reynolek) [inline packages/openapi-2-kong/src/declarative-config/__tests__/services.test.js]: To my knowledge we dont support `strip_path` at the root. I believe we should only support it at the route definition level (or at the `x-kong-route-default` level)

**Review by reynolek — CHANGES_REQUESTED** [review]

File: packages/openapi-2-kong/src/declarative-config/__tests__/services.test.js (hunk 4 above)
Comment (dimitropoulos) [reply packages/openapi-2-kong/src/declarative-config/__tests__/services.test.js]: the linear ticket and the jira ticket got crossed in my mind, sorry, fixed in 6ad54dfe5c6fb799919e05b1b5de74bfc341587f

File: packages/openapi-2-kong/src/declarative-config/__tests__/services.test.js (hunk 3 above)
Comment (dimitropoulos) [reply packages/openapi-2-kong/src/declarative-config/__tests__/services.test.js]: totally: was a mistake: fixed in b92d702bd1eb8739017b1b3cc4a486c1dc560327

File: packages/openapi-2-kong/src/declarative-config/__tests__/services.test.js (hunk 2 above)
Comment (dimitropoulos) [reply packages/openapi-2-kong/src/declarative-config/__tests__/services.test.js]: fixed in b92d702bd1eb8739017b1b3cc4a486c1dc560327

File: packages/openapi-2-kong/src/declarative-config/services.js (hunk 1 above)
Comment (dimitropoulos) [reply packages/openapi-2-kong/src/declarative-config/services.js]: fixed in 64498f939c28530bdc61712c850bd1a396408e24

**Review by reynolek — APPROVED** [review]

File: packages/openapi-2-kong/src/declarative-config/__tests__/services.test.js [hunk 5]
```diff
@@ -2,92 +2,85 @@
...
+
+const baseSpecResult = {
+  name: 'My_API',
+  url: 'https://server1.com/path',
+  plugins: [],
+  tags: ['Tag'],
+  routes: [
+    {
+      name: 'My_API-Cat_stuff-post',
+      strip_path: false,
```
Comment (Tieske) [inline packages/openapi-2-kong/src/declarative-config/__tests__/services.test.js]: I don't think this is wise. `strip_path` can be specified in `x-kong-route-defaults` and if not specified at all it should simply pick up Kong defaults when synced to Kong by decK.

Adding those named properties here will create a tightly coupled implementation. We should try to keep that to a minimum. So only where actual logic is involved, eg. with the request-validator generation.

dwyt @develohpanda @dimitropoulos ?

File: packages/openapi-2-kong/src/declarative-config/__tests__/services.test.js [hunk 6]
```diff
@@ -97,19 +90,119 @@ describe('services', () => {
...
+            ...baseSpecResult,
+            routes: [...baseSpecResult.routes].map(route => ({
+              ...route,
+              foo: 'bar',
+            })),
+          },
+        ]);
+      });
+
+      it('x-kong-route-defaults only works at the root', async () => {
```
Comment (Tieske) [inline packages/openapi-2-kong/src/declarative-config/__tests__/services.test.js]: To make it consistent with the plugins, I think this one should be possible on doc/path/operation level. Where the more specific level override the more generic ones (completely, so no merging of objects).

That would be consistent with the plugins, right @develohpanda ?

File: packages/openapi-2-kong/src/declarative-config/__tests__/services.test.js [hunk 7]
```diff
@@ -97,19 +90,119 @@ describe('services', () => {
...
+                foo: 'will be ignored, by design',
+              },
+            },
+          },
+        });
+
+        expect(generateServices(api, ['Tag'])).toEqual([baseSpecResult]);
+      });
+
+      it('allows overriding strip_path at the path level', async () => {
```
Comment (Tieske) [inline packages/openapi-2-kong/src/declarative-config/__tests__/services.test.js:438]: only allowing `strip_path` to cascade through the levels doesn't make sense to me, and I think it also complicates the implementation (though that's hard for me to judge)

File: packages/openapi-2-kong/src/declarative-config/__tests__/services.test.js (hunk 6 above)
Comment (develohpanda) [reply packages/openapi-2-kong/src/declarative-config/__tests__/services.test.js]: Yis that is how plugins behave now - no merging, most specific is used.

https://github.com/Kong/insomnia/blob/70738103d792ca7340464b585cad52842ca7be85/packages/openapi-2-kong/src/declarative-config/plugins.js#L179-L180

File: packages/openapi-2-kong/src/declarative-config/__tests__/services.test.js (hunk 5 above)
Comment (dimitropoulos) [reply packages/openapi-2-kong/src/declarative-config/__tests__/services.test.js]: I strongly agree with you.  The trouble is that the prior implementation hardcoded in `strip_path` to `false` and there was no way to specify it or not.  To change this behavior now would be a breaking change, we feel.

https://github.com/Kong/insomnia/blob/develop/packages/openapi-2-kong/src/declarative-config/services.js#L85

File: packages/openapi-2-kong/src/declarative-config/__tests__/services.test.js (hunk 6 above)
Comment (dimitropoulos) [reply packages/openapi-2-kong/src/declarative-config/__tests__/services.test.js]: I implemented it this way originally, and it was decided against because it kinda defeats the semantics of a top-level default, which is different from other properties.  @reynolek can say more, but that was the general idea.

File: packages/openapi-2-kong/src/declarative-config/__tests__/services.test.js (hunk 7 above)
Comment (dimitropoulos) [reply packages/openapi-2-kong/src/declarative-config/__tests__/services.test.js:438]: I agree, but [as I said above](https://github.com/Kong/insomnia/pull/3273#discussion_r618484885), the issue ultimately is that we had this hardcoded to false for all paths before which kinda locks us into this.

File: packages/openapi-2-kong/src/declarative-config/__tests__/services.test.js (hunk 6 above)
Comment (develohpanda) [reply packages/openapi-2-kong/src/declarative-config/__tests__/services.test.js]: FWIW I sort of agree with Thijs here about specificity. Use the top-level default but override when set at a more specific location

Comment (reynolek) [reply packages/openapi-2-kong/src/declarative-config/__tests__/services.test.js]: Lets go with making it consistent with other plugins. Looks like I directed you away from that earlier, not realizing we did it other ways in other places.

File: packages/openapi-2-kong/src/declarative-config/__tests__/services.test.js (hunk 5 above)
Comment (Tieske) [reply packages/openapi-2-kong/src/declarative-config/__tests__/services.test.js]: actually I was wrong. It should be set to `false`, if not specified in `x-kong-route-defaults`, only if `x-kong-route-defaults` sets it to `true` it should be `true`, in all other cases it should be `false`.

Reason being; the part of the path that is matched by Kong, will be stripped. This is because Kong treats it as a path-prefix. But the nature of OpenAPI specs is that the entire path is specified, not just a prefix. So if stripped, it would mean that the entire path is stripped, and hence every endpoint would be reduced to only `/`, which obviously is useless.

File: packages/openapi-2-kong/src/declarative-config/__tests__/services.test.js (hunk 6 above)
Comment (Tieske) [reply packages/openapi-2-kong/src/declarative-config/__tests__/services.test.js]: agreed, merging objects can have weird edge cases. Let's put the can of worms back on the shelf 😄

File: packages/openapi-2-kong/src/declarative-config/__tests__/services.test.js (hunk 7 above)
Comment (Tieske) [reply packages/openapi-2-kong/src/declarative-config/__tests__/services.test.js:438]: yep, should be as per other comments, full set of defaults, no merging. And strip_path only goes to `true` if there is a `true` in the defaults, if absent it goes to `false`.

Comment (dimitropoulos) [reply packages/openapi-2-kong/src/declarative-config/__tests__/services.test.js:438]: should be updated now.  enough has changed anyway that I think another top-level look is necessitated.

File: packages/openapi-2-kong/src/declarative-config/__tests__/services.test.js (hunk 5 above)
Comment (dimitropoulos) [reply packages/openapi-2-kong/src/declarative-config/__tests__/services.test.js]: should be updated now.  enough has changed anyway that I think another top-level look is necessitated.

File: packages/openapi-2-kong/src/declarative-config/__tests__/services.test.js (hunk 6 above)
Comment (dimitropoulos) [reply packages/openapi-2-kong/src/declarative-config/__tests__/services.test.js]: should be updated now.  enough has changed anyway that I think another top-level look is necessitated.

File: packages/openapi-2-kong/src/declarative-config/services.js [hunk 8]
```diff
@@ -81,16 +82,34 @@ export function generateService(
         continue;
       }
 
+      // $FlowFixMe
+      const operationRouteDefaults = pathItem['x-kong-route-defaults'];
```
Comment (Tieske) [inline packages/openapi-2-kong/src/declarative-config/services.js]: This is actually the "path" default

File: packages/openapi-2-kong/src/declarative-config/services.js [hunk 9]
```diff
@@ -81,16 +82,34 @@ export function generateService(
...
+      const pathRouteDefaults = api.paths['x-kong-route-defaults'];
```
Comment (Tieske) [inline packages/openapi-2-kong/src/declarative-config/services.js]: this is not right, a path default on that array does not make sense. The array has a 1-on-1 relation ship with the root document, hence from a scoping perspective this would be identical to `x-kong-route-defaults` on root level.

**Review by Tieske — CHANGES_REQUESTED** [review]
I think we should close this and replace it with #3319

File: packages/openapi-2-kong/src/declarative-config/services.js [hunk 10]
```diff
@@ -42,9 +42,18 @@ export function generateService(
...
+  const serviceDefaults = api['x-kong-service-defaults'] || {};
+  if (typeof serviceDefaults !== 'object') {
+    throw new Error(`expected 'x-kong-service-defaults' to be an object`);
+  }
+
   const service: DCService = {
+    ...serviceDefaults,
     name,
-    // remove semicolon ie. convert https: to https
+    // remove semicolon i.e. convert `https` to `https`
```
Comment (Tieske) [inline packages/openapi-2-kong/src/declarative-config/services.js]:
```suggestion
    // remove semicolon i.e. convert `https:` to `https`
```

**Review by Tieske — APPROVED** [review]
